Minor translator cleanups

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Cleanups, comment.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
dl/t/sm: Always treat uninitialized last_translated_timestamp as nullopt

Previously would use timestamp::min() as a sort of proxy value for 'unset',
but that's not really needed.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dl/t/ds: data_source::{update->replicate}_highest_translated_offset

Renamed to avoid confusion with a plain old update function in translation_stm

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
dl/t: Add a clarifying comment in translation/deps

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

andrwng commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -210,7 +210,7 @@ ss::future<> translation_stm::apply_raft_snapshot(const iobuf&) {
     vlog(_log.debug, "Applying raft snapshot, resetting state");
     _highest_translated_offset = kafka::offset{};
-    _last_translated_timestamp = model::timestamp::min();
+    _last_translated_timestamp = std::nullopt;
```
There's also `timestamp::min()` in `do_apply()` that should probably be removed?

oleiman replied:
that one is intentional - if `value.last_translated_timestamp.has_value()`, then its value must be `>= timestamp::min()`, so we can just resolve the cached copy to `timestamp::min()` to make it work w/ `std::max`. 

though i just remembered that `std::optional` supports comparison operators, so I believe this would have the same effect:

```c++
_last_translated_timestamp = std::max(_last_translated_timestamp, value.last_translated_timestamp);
```

in the sense of always blowing away the cached copy if it's nullopt and comparing the two otherwise.

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -271,8 +271,9 @@ ss::future<> partition_translator::translate_until_stopped() {
 
-        reset_error = co_await _data_source->update_highest_translated_offset(
-          last_translated_offset, _term, wait_timeout, _as);
+        reset_error
+          = co_await _data_source->replicate_highest_translated_offset(
+            last_translated_offset, _term, wait_timeout, _as);
```
For posterity, we discussed offline that maybe this should only update the lag timestamp if we are replicating up to the translation target. I don't think the dust settled on that conversation, but thought I'd mention it here

oleiman replied:
ya, i will tack it onto this diff so we can see it _in situ_

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62610-->
<!--a884d115-cd80-4d0d-a28a-2cccc1109dfe-->
<!--title-->
## Retry command for [Build#62610](https://buildkite.com/redpanda/redpanda/builds/62610)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/cluster_restore_test.py::DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq@{"catalog_type":"rest_hadoop","cloud_storage_type":1}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--f8fe44c4-df60-4f90-98b6-3ee4a58809bd-->
<!--TEST RESULTS-->
<!--32ed2f30-329b-4bf2-9a40-325ec0dd7a32-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62610">build#62610</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62610#01956286-270d-4a47-974e-dd8a9249e69c | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562ca-397c-4cc0-b940-3f5d44b31134 | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562e5-9750-4938-9edc-94da4787656e | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562e5-9751-46e2-b73f-d5880fa9d328 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562ca-397e-41c6-aa4e-1869b8ed1920 | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562ca-397f-4938-bfa5-333621f9ffe7 | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562e5-974f-4b41-87d4-a70289943884 | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562ca-397d-435f-96a7-b0831d43ea88 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562e5-9751-46e2-b73f-d5880fa9d328 | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562ca-397e-41c6-aa4e-1869b8ed1920 | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562e5-974f-4d15-887e-fe79f122af28 | FLAKY | 1/5 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562ca-397f-4938-bfa5-333621f9ffe7 | FAIL | 0/20 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562ca-397d-435f-96a7-b0831d43ea88 | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562e5-9751-46e2-b73f-d5880fa9d328 | FLAKY | 1/8 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562ca-397e-41c6-aa4e-1869b8ed1920 | FLAKY | 1/8 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562e5-974f-4b41-87d4-a70289943884 | FLAKY | 1/6 |
| rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562e5-9750-4938-9edc-94da4787656e | FLAKY | 1/2 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562e5-974f-4b41-87d4-a70289943884 | FLAKY | 1/2 |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=True.allow_gaps_cluster_level=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/62610#019562e5-974f-4b41-87d4-a70289943884 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62637">build#62637</a></summary>


| test_id                                                                                                                                                                               | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195647a-4107-42a7-87ec-9c2b36a5f849 | FLAKY         | 1/2      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195648f-a890-4930-a113-9b616473f841 | FLAKY         | 1/2      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195647a-4109-403b-9d7f-bbdb9f7f27ea | FLAKY         | 1/4      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195648f-a88f-4701-afce-d038883457fc | FLAKY         | 1/2      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195648f-a890-40dc-a974-d73ed94a0c66 | FLAKY         | 1/4      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195647a-4107-42a7-87ec-9c2b36a5f849 | FLAKY         | 1/6      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195647a-4108-4545-871b-74d97ae4fd05 | FLAKY         | 1/3      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195647a-4109-403b-9d7f-bbdb9f7f27ea | FLAKY         | 1/2      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195648f-a88f-4701-afce-d038883457fc | FLAKY         | 1/16     |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195648f-a890-40dc-a974-d73ed94a0c66 | FLAKY         | 1/2      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195648f-a890-4930-a113-9b616473f841 | FLAKY         | 1/3      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195647a-4108-4545-871b-74d97ae4fd05 | FLAKY         | 1/2      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195648f-a891-4571-8f64-9fe6d65e33ac | FLAKY         | 1/2      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195647a-4109-439c-a67c-54b3e031eba2 | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195647a-4108-4545-871b-74d97ae4fd05 | FLAKY         | 1/2      |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=None.allow_gaps_cluster_level=True                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195647a-4109-403b-9d7f-bbdb9f7f27ea | FLAKY         | 1/2      |
| rptest.tests.write_caching_fi_e2e_test.WriteCachingFailureInjectionE2ETest.test_crash_all_with_consumer_group                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62637#0195648f-a88f-4701-afce-d038883457fc | FLAKY         | 1/2      |
</details>
<!--/body-->

andrwng commented:
> ## CI test results
> test results on [build#62610](https://buildkite.com/redpanda/redpanda/builds/62610)

Wow that's a lot of failures of the cluster restore test. I'll take a look

oleiman commented:
> > ## CI test results
> > test results on [build#62610](https://buildkite.com/redpanda/redpanda/builds/62610)
> 
> Wow that's a lot of failures of the cluster restore test. I'll take a look

cool, thanks. `rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP` in particular passed 0/20 where the others were flaky. reproduces locally fwiw.

oleiman committed:
dl/t: Don't update _translation_target until we reach the previous one

This should prevent us from getting in a situation where we're continually
chasing the tip of the log, never catching it, and so never updating our the
replicated last timestamp, which we want to use for approximating lag.
